### ISSUE TITLE
New client/server architecture

### ISSUE THREAD

**Issue by luizribeiro** [luizribeiro/labby#4]
Probably something like this:

```
┌ Client Interfaces ─ ─                    ┌ Main Process─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐
                       │                                                                                  
│  ┌────────────────┐                      │                                                             │
   │  Command Line  │◀─┼─────────┐            ┌─────────────────────┐           ┌─────────────────────┐   
│  └────────────────┘            │         │  │  Server Controller  │  pub/sub  │  Experiment Runner  │  │
                       │         ├───────────▶│       Thread        │◀─────────▶│       Thread        │   
│  ┌────────────────┐            │         │  └─────────────────────┘           └─────────────────────┘  │
   │    Jupyter     │◀─┼─────────┘                       ▲                                 ▲              
│  └────────────────┘                      │             │                                 │             │
                       │   nng + msgpack                 └────────────────┬────────────────┘              
│                         (request/reply)  │                              │                              │
                       │                                                  │   priority queue              
│                                          │                              ▼                              │
                       │                     ┌────────────────────────────────────────────────────────┐   
│                                          │ │             Serial Port Controller Thread              │  │
                       │                     └────────────────────────────────────────────────────────┘   
│                                          │                                                             │
                       │                                                                                  
└ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─                    └ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘
```

* **Server Controller Thread:**
  * Keeps track of running experiments
  * Keeps track of devices
  * Answers questions from clients about state
    * State of all devices
    * Information about a specific device
    * Information about running experiments
  * Answers commands from clients
    * Start/stop experiments
  * Instantiates experiments and subscribes to updates from them
* **Experiment Runner Thread:**
  * Loops until experiment is done
  * Basically just runs ExperimentRunner and waits until it's done
* **Serial Port Controller Thread:**
  * One per serial port
  * Keeps a priority queue of messages
  * Messages from experiments have the highest priority

**Comment by luizribeiro** [comment]
This has been implemented 🎉
